### PR TITLE
Port libosmium examples

### PR DESCRIPTION
**Pull request by boxdot**
Very simple examples

* [x] Read
* [x] Count
* [x] Debug
* [ ] Tiles

Still reasonably simple examples

* [ ] amenity_list
* [ ] read_with_progress
* [ ] filter_discussions
* [ ] convert
* [x] pub_names
* [x] road_length

### CODE REVIEW

File: examples/debug.rs [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+}
+
+impl fmt::Display for FixedI64 {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let value: f64 = self.value();
+        write!(f, "{}", value)
+    }
+}
+
+impl fmt::Debug for FixedI64 {
```
Comment (VeaaC) [inline examples/debug.rs]: Do you need both for this example?

File: examples/debug.rs [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+
+    if types.contains('w') {
+        for way in archive.ways() {
+            let tags: Result<Vec<_>, _> = osmflat::tags(&archive, way.tags()).collect();
+            let way = Way {
+                id: way.id(),
+                tags: tags?,
+                nodes: way
+                    .refs()
+                    .map(|idx| nodes_index.at(idx as usize).value())
```
Comment (VeaaC) [inline examples/debug.rs]: nodes_index.slice(way.refs().start as usize..way.refs().end as usize)?

PS: We could enhance flatdata to allow indexing by non-usize Ranges

File: examples/debug.rs [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+            source: strings.substring(header.source_idx() as usize)?,
+            osmosis_replication_timestamp: header.osmosis_replication_timestamp(),
+            osmosis_replication_sequence_number: header.osmosis_replication_sequence_number(),
+            osmosis_replication_base_url: strings
+                .substring(header.osmosis_replication_base_url_idx() as usize)?,
+        })
+    }
+}
+
+#[derive(Debug)]
```
Comment (VeaaC) [inline examples/debug.rs]: PS: Are all those debug structures, and their conversion function maybe much longer than just printing directly the flatdata structure manually? Brevity might be useful here.

**Review by VeaaC — APPROVED** [review]

File: examples/debug.rs (hunk 1 above)
Comment (boxdot) [reply examples/debug.rs]: Debug, yes, but not Display. Removed.

File: examples/debug.rs (hunk 3 above)
Comment (boxdot) [reply examples/debug.rs]: Yeah, I am also undecided about that. Somehow printing manually feels weird. Might be useful to learn how to wrap data, but I am not sure.

File: examples/debug.rs (hunk 2 above)
Comment (boxdot) [reply examples/debug.rs]: Would still need to call `map` to get the value.

File: examples/road_length.rs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    let lengths = highways.map(|way| {
+        let coords = way.refs().map(|idx| {
+            // A way references a range of nodes by storing a contiguous range of
+            // indexes in `nodes_index`. Each of these references a node in `nodes`.
+            Coords::from_node(nodes.at(nodes_index.at(idx as usize).value() as usize))
+        });
+        let length: f64 = coords
+            .clone()
+            .zip(coords.skip(1))
+            .map(|(from, to)| haversine_distance(from, to))
```
Comment (VeaaC) [inline examples/road_length.rs]: just pass haversine_distance to map directly

File: src/lib.rs [hunk 5]
```diff
@@ -24,3 +25,89 @@ pub fn tags<'a>(
         Ok((key, val))
     })
 }
+
+/// Helper function to iterate through tags from osmflat.
+#[inline]
+pub fn tags_raw<'a>(
```
Comment (VeaaC) [inline src/lib.rs]: Remove all the non-raw versions (also from flatdata)?

Comment (boxdot) [reply src/lib.rs]: Done.

**Comment by boxdot** [comment]
@VeaaC Could you please take a look again. I think I am ready to merge this.

**Pull request merged**